File: src/bridge.cpp
#include "webview_official/src/bridge.h"
#include "webview_official/src/bridge.rs.h"

#include <WebView2.h>
#include <WebView2EnvironmentOptions.h>

#include <windows.h>
#include <wrl.h>

#include <algorithm>
#include <iterator>
#include <stdexcept>
#include <string>
#include <string_view>

using SPWSTR = std::unique_ptr<wchar_t[], decltype(&::CoTaskMemFree)>;

std::wstring to_wstring(const rust::Vec<uint16_t> source)
{
    return {reinterpret_cast<const wchar_t *>(source.data()), source.size()};
}

std::wstring to_wstring(rust::Slice<const uint16_t> source)
{
    return {reinterpret_cast<const wchar_t *>(source.data()), source.size()};
}

rust::Vec<uint16_t> to_vec(std::wstring_view source)
{
    rust::Vec<uint16_t> result;

    result.reserve(source.size());
    std::transform(source.begin(), source.end(), std::back_inserter(result), [](wchar_t ch) {
        return static_cast<uint16_t>(ch);
    });

    return result;
}

class WebView2Environment::impl
{
public:
    impl();
    ~impl();

    void CreateWebView2Controller(ptrdiff_t parent_window, rust::Box<CreateWebView2ControllerCompletedHandler> handler) const;

    Microsoft::WRL::ComPtr<ICoreWebView2Environment> m_environment;
};

class WebView2Controller::impl
{
public:
    Microsoft::WRL::ComPtr<ICoreWebView2Controller> m_controller;
};

class WebView2::impl
{
public:
    Microsoft::WRL::ComPtr<ICoreWebView2> m_webview;
};

WebView2Environment::impl::impl()
{
}

WebView2Environment::impl::~impl()
{
}

void WebView2Environment::impl::CreateWebView2Controller(ptrdiff_t parent_window, rust::Box<CreateWebView2ControllerCompletedHandler> handler) const
{
    if (!m_environment)
    {
        throw std::runtime_error("ICoreWebView2Environment creation failed");
    }

    auto callback = Microsoft::WRL::Callback<ICoreWebView2CreateCoreWebView2ControllerCompletedHandler>(
        [handler = std::move(handler)](HRESULT hr, ICoreWebView2Controller *controller) mutable noexcept {
            std::unique_ptr<WebView2Controller> result;

            if (SUCCEEDED(hr) && nullptr != controller)
            {
                result = std::make_unique<WebView2Controller>();
                result->m_pimpl->m_controller = controller;
            }

            invoke_controller_complete(std::move(handler), std::move(result));
            return S_OK;
        });

    if (FAILED(m_environment->CreateCoreWebView2Controller(reinterpret_cast<HWND>(parent_window), callback.Get())))
    {
        throw std::runtime_error("CreateCoreWebView2Controller failed");
    }
}

void new_webview2_environment(rust::Box<CreateWebView2EnvironmentCompletedHandler> handler)
{
    auto callback = Microsoft::WRL::Callback<ICoreWebView2CreateCoreWebView2EnvironmentCompletedHandler>(
        [handler = std::move(handler)](HRESULT hr, ICoreWebView2Environment *environment) mutable noexcept {
            std::unique_ptr<WebView2Environment> result;

            if (SUCCEEDED(hr) && nullptr != environment)
            {
                result = std::make_unique<WebView2Environment>();
                result->m_pimpl->m_environment = environment;
            }

            invoke_environment_complete(std::move(handler), std::move(result));
            return S_OK;
        });

    if (FAILED(CreateCoreWebView2Environment(callback.Get())))
    {
        throw std::runtime_error("CreateCoreWebView2Environment failed");
    }
}

void new_webview2_environment_with_options(rust::Slice<const uint16_t> browser_executable_folder,
                                           rust::Slice<const uint16_t> user_data_folder,
                                           const WebView2EnvironmentOptions &options,
                                           rust::Box<CreateWebView2EnvironmentCompletedHandler> handler)
{
    auto browserExecutableFolder = to_wstring(browser_executable_folder);
    auto userDataFolder = to_wstring(user_data_folder);
    auto spOptions = Microsoft::WRL::Make<CoreWebView2EnvironmentOptions>();

    if (!options.aditional_browser_arguments.empty())
    {
        auto additionalBrowserArguments = to_wstring(options.aditional_browser_arguments);

        if (FAILED(spOptions->put_AdditionalBrowserArguments(additionalBrowserArguments.c_str())))
        {
            throw std::runtime_error("put_AdditionalBrowserArguments failed");
        }
    }

    if (!options.language.empty())
    {
        auto language = to_wstring(options.language);

        if (FAILED(spOptions->put_Language(language.c_str())))
        {
            throw std::runtime_error("put_Language failed");
        }
    }

    if (!options.target_compatible_browser_version.empty())
    {
        auto targetCompatibleBrowserVersion = to_wstring(options.target_compatible_browser_version);

        if (FAILED(spOptions->put_TargetCompatibleBrowserVersion(targetCompatibleBrowserVersion.c_str())))
        {
            throw std::runtime_error("put_TargetCompatibleBrowserVersion failed");
        }
    }

    if (FAILED(spOptions->put_AllowSingleSignOnUsingOSPrimaryAccount(options.allow_single_sign_on_using_os_primary_account)))
    {
        throw std::runtime_error("put_AllowSingleSignOnUsingOSPrimaryAccount failed");
    }

    auto callback = Microsoft::WRL::Callback<ICoreWebView2CreateCoreWebView2EnvironmentCompletedHandler>(
        [handler = std::move(handler)](HRESULT hr, ICoreWebView2Environment *environment) mutable noexcept {
            std::unique_ptr<WebView2Environment> result;

            if (SUCCEEDED(hr) && nullptr != environment)
            {
                result = std::make_unique<WebView2Environment>();
                result->m_pimpl->m_environment = environment;
            }

            invoke_environment_complete(std::move(handler), std::move(result));
            return S_OK;
        });

    if (FAILED(CreateCoreWebView2EnvironmentWithOptions(browserExecutableFolder.c_str(),
                                                        userDataFolder.c_str(),
                                                        spOptions.Get(),
                                                        callback.Get())))
    {
        throw std::runtime_error("CreateCoreWebView2Environment failed");
    }
}

rust::Vec<uint16_t> get_available_webview2_browser_version_string(rust::Slice<const uint16_t> browser_executable_folder)
{
    std::wstring browserExecutableFolder{
        reinterpret_cast<const wchar_t *>(browser_executable_folder.data()),
        browser_executable_folder.size()};
    PWSTR version = nullptr;

    if (FAILED(GetAvailableCoreWebView2BrowserVersionString(browserExecutableFolder.c_str(), &version)))
    {
        throw std::runtime_error("GetAvailableCoreWebView2BrowserVersionString failed");
    }

    SPWSTR cleanup{version, ::CoTaskMemFree};

    return to_vec(version);
}

int8_t compare_browser_versions(rust::Slice<const uint16_t> version1, rust::Slice<const uint16_t> version2)
{
    int result = 0;

    if (FAILED(CompareBrowserVersions(to_wstring(version1).c_str(), to_wstring(version2).c_str(), &result)))
    {
        throw std::runtime_error("CompareBrowserVersions failed");
    }

    if (result > 0)
    {
        return 1;
    }
    else if (result < 0)
    {
        return -1;
    }

    return 0;
}

WebView2Environment::WebView2Environment()
    : m_pimpl{std::make_unique<impl>()}
{
}

WebView2Environment::~WebView2Environment()
{
}

void WebView2Environment::create_webview2_controller(ptrdiff_t parent_window, rust::Box<CreateWebView2ControllerCompletedHandler> handler) const
{
    m_pimpl->CreateWebView2Controller(parent_window, std::move(handler));
}

WebView2Controller::WebView2Controller()
    : m_pimpl{std::make_unique<impl>()}
{
}

WebView2Controller::~WebView2Controller()
{
}

WebView2::WebView2()
    : m_pimpl{std::make_unique<impl>()}
{
}

WebView2::~WebView2()
{
}

File: src/bridge.rs
#[cxx::bridge]
pub mod core {
    #[derive(Debug, Default)]
    struct WebView2EnvironmentOptions {
        aditional_browser_arguments: Vec<u16>,
        language: Vec<u16>,
        target_compatible_browser_version: Vec<u16>,
        allow_single_sign_on_using_os_primary_account: bool,
    }

    #[derive(Debug, Default)]
    struct WebView2ControllerBounds {
        left: i32,
        top: i32,
        right: i32,
        bottom: i32,
    }

    #[derive(Debug)]
    struct WebView2Settings {
        is_script_enabled: bool,
        is_web_message_enabled: bool,
        are_default_script_dialogs_enabled: bool,
        is_status_bar_enabled: bool,
        are_dev_tools_enabled: bool,
        are_default_context_menus_enabled: bool,
        is_zoom_control_enabled: bool,
        is_built_in_error_page_enabled: bool,
    }

    extern "Rust" {
        fn to_utf16(value: &str) -> Vec<u16>;
        fn from_utf16(value: &[u16]) -> String;

        type CreateWebView2EnvironmentCompletedHandler;

        fn invoke_environment_complete(
            handler: Box<CreateWebView2EnvironmentCompletedHandler>,
            environment: SharedPtr<WebView2Environment>,
        );

        type CreateWebView2ControllerCompletedHandler;

        fn invoke_controller_complete(
            handler: Box<CreateWebView2ControllerCompletedHandler>,
            controller: SharedPtr<WebView2Controller>,
        );

        type NavigationCompletedHandler;

        fn invoke_navigation_complete(handler: Box<NavigationCompletedHandler>, webview: &WebView2);

        type AddScriptToExecuteOnDocumentCreatedCompletedHandler;

        fn invoke_add_script_on_document_created_complete(
            handler: Box<AddScriptToExecuteOnDocumentCreatedCompletedHandler>,
            id: Vec<u16>,
        );

        type ExecuteScriptCompletedHandler;

        fn invoke_execute_script_complete(
            handler: Box<ExecuteScriptCompletedHandler>,
            result: Vec<u16>,
        );

        type WebMessageReceivedHandler;

        fn invoke_web_message_received(
            handler: &WebMessageReceivedHandler,
            source: Vec<u16>,
            message: Vec<u16>,
        );
    }

    unsafe extern "C++" {
        include!("webview_official/src/bridge.h");

        fn new_webview2_environment(
            handler: Box<CreateWebView2EnvironmentCompletedHandler>,
        ) -> Result<()>;

        fn new_webview2_environment_with_options(
            browser_executable_folder: &[u16],
            user_data_folder: &[u16],
            options: &WebView2EnvironmentOptions,
            handler: Box<CreateWebView2EnvironmentCompletedHandler>,
        ) -> Result<()>;

        fn get_available_webview2_browser_version_string(
            browser_executable_folder: &[u16],
        ) -> Result<Vec<u16>>;

        fn compare_browser_versions(version1: &[u16], version2: &[u16]) -> Result<i8>;

        type WebView2Environment;

        fn create_webview2_controller(
            self: &WebView2Environment,
            parent_window: isize,
            handler: Box<CreateWebView2ControllerCompletedHandler>,
        ) -> Result<&WebView2Environment>;

        type WebView2Controller;

        fn visible(self: &WebView2Controller, value: bool) -> Result<&WebView2Controller>;
        fn get_visible(self: &WebView2Controller) -> Result<bool>;
        fn bounds(
            self: &WebView2Controller,
            value: WebView2ControllerBounds,
        ) -> Result<&WebView2Controller>;
        fn get_bounds(self: &WebView2Controller) -> Result<WebView2ControllerBounds>;
        fn close(self: &WebView2Controller) -> Result<()>;
        fn get_webview(self: &WebView2Controller) -> Result<SharedPtr<WebView2>>;

        type WebView2;

        fn settings(self: &WebView2, value: WebView2Settings) -> Result<&WebView2>;
        fn get_settings(self: &WebView2) -> Result<WebView2Settings>;
        fn navigate(
            self: &WebView2,
            url: &[u16],
            handler: Box<NavigationCompletedHandler>,
        ) -> Result<&WebView2>;
        fn navigate_to_string(
            self: &WebView2,
            html_content: &[u16],
            handler: Box<NavigationCompletedHandler>,
        ) -> Result<&WebView2>;
        fn add_script_to_execute_on_document_created(
            self: &WebView2,
            javascript: &[u16],
            handler: Box<AddScriptToExecuteOnDocumentCreatedCompletedHandler>,
        ) -> Result<&WebView2>;
        fn remove_script_to_execute_on_document_created(
            self: &WebView2,
            id: &[u16],
        ) -> Result<&WebView2>;
        fn execute_script(
            self: &WebView2,
            javascript: &[u16],
            handler: Box<ExecuteScriptCompletedHandler>,
        ) -> Result<&WebView2>;
        fn reload(self: &WebView2) -> Result<&WebView2>;
        fn post_web_message(self: &WebView2, json_message: &[u16]) -> Result<&WebView2>;
        fn add_web_message_received(
            self: &WebView2,
            handler: Box<WebMessageReceivedHandler>,
        ) -> Result<i64>;
        fn remove_web_message_received(self: &WebView2, token: i64) -> Result<&WebView2>;
        fn stop(self: &WebView2) -> Result<&WebView2>;
        fn get_document_title(self: &WebView2) -> Result<Vec<u16>>;
        fn open_dev_tools_window(self: &WebView2) -> Result<&WebView2>;
    }
}

pub fn to_utf16(value: &str) -> Vec<u16> {
    value.encode_utf16().collect()
}

pub fn from_utf16(value: &[u16]) -> String {
    match String::from_utf16(value) {
        Ok(result) => result,
        Err(_) => String::new(),
    }
}

impl core::WebView2EnvironmentOptions {
    pub fn new(
        aditional_browser_arguments: &str,
        language: &str,
        target_compatible_browser_version: &str,
        allow_single_sign_on_using_os_primary_account: bool,
    ) -> core::WebView2EnvironmentOptions {
        core::WebView2EnvironmentOptions {
            aditional_browser_arguments: to_utf16(aditional_browser_arguments),
            language: to_utf16(language),
            target_compatible_browser_version: to_utf16(target_compatible_browser_version),
            allow_single_sign_on_using_os_primary_account,
        }
    }
}

type EnvironmentCompletedCallback = Box<dyn FnOnce(cxx::SharedPtr<core::WebView2Environment>)>;

pub struct CreateWebView2EnvironmentCompletedHandler {
    callback: EnvironmentCompletedCallback,
}

impl CreateWebView2EnvironmentCompletedHandler {
    pub fn new(callback: EnvironmentCompletedCallback) -> Self {
        Self { callback }
    }

    pub fn handle(self, environment: cxx::SharedPtr<core::WebView2Environment>) {
        (self.callback)(environment);
    }
}

pub fn invoke_environment_complete(
    handler: Box<CreateWebView2EnvironmentCompletedHandler>,
    environment: cxx::SharedPtr<core::WebView2Environment>,
) {
    handler.handle(environment);
}

type ControllerCompletedCallback = Box<dyn FnOnce(cxx::SharedPtr<core::WebView2Controller>)>;

pub struct CreateWebView2ControllerCompletedHandler {
    callback: ControllerCompletedCallback,
}

impl CreateWebView2ControllerCompletedHandler {
    pub fn new(callback: ControllerCompletedCallback) -> Self {
        Self { callback }
    }

    pub fn handle(self, controller: cxx::SharedPtr<core::WebView2Controller>) {
        (self.callback)(controller);
    }
}

pub fn invoke_controller_complete(
    handler: Box<CreateWebView2ControllerCompletedHandler>,
    controller: cxx::SharedPtr<core::WebView2Controller>,
) {
    handler.handle(controller);
}

type NavigationCompletedCallback = Box<dyn FnOnce(&core::WebView2)>;

pub struct NavigationCompletedHandler {
    callback: NavigationCompletedCallback,
}

impl NavigationCompletedHandler {
    pub fn new(callback: NavigationCompletedCallback) -> Self {
        Self { callback }
    }

    pub fn handle(self, webview: &core::WebView2) {
        (self.callback)(webview);
    }
}

pub fn invoke_navigation_complete(
    handler: Box<NavigationCompletedHandler>,
    webview: &core::WebView2,
) {
    handler.handle(webview)
}

type AddScriptToExecuteOnDocumentCreatedCompletedCallback = Box<dyn FnOnce(String)>;

pub struct AddScriptToExecuteOnDocumentCreatedCompletedHandler {
    callback: ExecuteScriptCompletedCallback,
}

impl AddScriptToExecuteOnDocumentCreatedCompletedHandler {
    pub fn new(callback: AddScriptToExecuteOnDocumentCreatedCompletedCallback) -> Self {
        Self { callback }
    }

    pub fn handle(self, result: String) {
        (self.callback)(result);
    }
}

pub fn invoke_add_script_on_document_created_complete(
    handler: Box<AddScriptToExecuteOnDocumentCreatedCompletedHandler>,
    id: Vec<u16>,
) {
    handler.handle(from_utf16(&id));
}

type ExecuteScriptCompletedCallback = Box<dyn FnOnce(String)>;

pub struct ExecuteScriptCompletedHandler {
    callback: ExecuteScriptCompletedCallback,
}

impl ExecuteScriptCompletedHandler {
    pub fn new(callback: ExecuteScriptCompletedCallback) -> Self {
        Self { callback }
    }

    pub fn handle(self, result: String) {
        (self.callback)(result);
    }
}

pub fn invoke_execute_script_complete(
    handler: Box<ExecuteScriptCompletedHandler>,
    result: Vec<u16>,
) {
    handler.handle(from_utf16(&result));
}

type WebMessageReceivedCallback = Box<dyn Fn(String, String)>;

pub struct WebMessageReceivedHandler {
    callback: WebMessageReceivedCallback,
}

impl WebMessageReceivedHandler {
    pub fn new(callback: WebMessageReceivedCallback) -> Self {
        Self { callback }
    }

    pub fn handle(&self, source: String, message: String) {
        (self.callback)(source, message);
    }
}

pub fn invoke_web_message_received(
    handler: &WebMessageReceivedHandler,
    source: Vec<u16>,
    message: Vec<u16>,
) {
    handler.handle(from_utf16(&source), from_utf16(&message));
}
